ros2_control_demo_hardware/src/rrbot_actuator.cpp

```diff
@@ -0,0 +1,178 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//
+// Authors: Subhas Das, Denis Stogl
+//
+
+#include "ros2_control_demo_hardware/rrbot_actuator.hpp"
+
+#include <chrono>
+#include <cmath>
+#include <limits>
+#include <memory>
+#include <vector>
+
+#include "hardware_interface/actuator_interface.hpp"
+#include "hardware_interface/types/hardware_interface_type_values.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+namespace ros2_control_demo_hardware
+{
+CallbackReturn RRBotModularJoint::on_init(const hardware_interface::HardwareInfo & info)
+{
+  if (hardware_interface::ActuatorInterface::on_init(info) != CallbackReturn::SUCCESS)
+  {
+    return CallbackReturn::ERROR;
+  }
+
+  hw_start_sec_ = stod(info_.hardware_parameters["example_param_hw_start_duration_sec"]);
+  hw_stop_sec_ = stod(info_.hardware_parameters["example_param_hw_stop_duration_sec"]);
+  hw_slowdown_ = stod(info_.hardware_parameters["example_param_hw_slowdown"]);
+
+  hw_joint_state_ = std::numeric_limits<double>::quiet_NaN();
+  hw_joint_command_ = std::numeric_limits<double>::quiet_NaN();
+
+  const hardware_interface::ComponentInfo & joint = info_.joints[0];
+  // RRBotModularJoint has exactly one state and command interface on each joint
+  if (joint.command_interfaces.size() != 1)
+  {
+    RCLCPP_FATAL(
+      rclcpp::get_logger("RRBotModularJoint"),
+      "Joint '%s' has %zu command interfaces found. 1 expected.", joint.name.c_str(),
+      joint.command_interfaces.size());
+    return CallbackReturn::ERROR;
+  }
+
+  if (joint.command_interfaces[0].name != hardware_interface::HW_IF_POSITION)
+  {
+    RCLCPP_FATAL(
+      rclcpp::get_logger("RRBotModularJoint"),
+      "Joint '%s' have %s command interfaces found. '%s' expected.", joint.name.c_str(),
+      joint.command_interfaces[0].name.c_str(), hardware_interface::HW_IF_POSITION);
+    return CallbackReturn::ERROR;
+  }
+
+  if (joint.state_interfaces.size() != 1)
+  {
+    RCLCPP_FATAL(
+      rclcpp::get_logger("RRBotModularJoint"), "Joint '%s' has %zu state interface. 1 expected.",
+      joint.name.c_str(), joint.state_interfaces.size());
+    return CallbackReturn::ERROR;
+  }
+
+  if (joint.state_interfaces[0].name != hardware_interface::HW_IF_POSITION)
+  {
+    RCLCPP_FATAL(
+      rclcpp::get_logger("RRBotModularJoint"), "Joint '%s' have %s state interface. '%s' expected.",
+      joint.name.c_str(), joint.state_interfaces[0].name.c_str(),
+      hardware_interface::HW_IF_POSITION);
+    return CallbackReturn::ERROR;
+  }
+
+  return CallbackReturn::SUCCESS;
+}
+
+std::vector<hardware_interface::StateInterface> RRBotModularJoint::export_state_interfaces()
+{
+  std::vector<hardware_interface::StateInterface> state_interfaces;
+
+  state_interfaces.emplace_back(hardware_interface::StateInterface(
+    info_.joints[0].name, hardware_interface::HW_IF_POSITION, &hw_joint_state_));
+
+  return state_interfaces;
```

Comment:
```suggestion
  return {hardware_interface::StateInterface(
    info_.joints[0].name, hardware_interface::HW_IF_POSITION, &hw_joint_state_)};
```

could something like this work?

Reply:
It could, but I don't know... It's example... having more, clearer code is not so bad idea I think.